### DISCUSSION TITLE
New module @turf/polygon-slice

### DISCUSSION THREAD

**Discussion by DenisCarriere** [Turfjs/turf#2782] [Feature proposals]
New module ~`@turf/slice`~ `@turf/polygon-slice`, currently on [**`slice`** branch](https://github.com/Turfjs/turf/tree/slice/packages/turf-polygon-slice)

Contributions and comments are welcomed before we publish to TurfJS.

**Example**
```javascript
var sliced = turf.polygonSlice(polygon, linestring);
```

**Current fallbacks of `PolyK`**
- Linestring is limited to two segments
- Coordinate precision is limited to a few decimal points.
- Output & Input is a very strange array [x1 y1, x2, y2, ...]

**JSDocs**
```javascript
/**
 * Takes a {@link Polygon} and cuts it with a {@link Linestring}. Note the linestring must be a straight line (eg made of only two points).
 * Properties from the input polygon will be retained on output polygons. Internally uses [polyK](http://polyk.ivank.net/) to perform split.
 *
 * @name slice
 * @param {Feature<Polygon>} poly - single Polygon Feature
 * @param {Feature<LineString>} line - single Polyline Feature
 * @return {FeatureCollection<Polygon>} A FeatureCollection of polygons
 */
```

**Input**

![image](https://cloud.githubusercontent.com/assets/550895/23344200/12550b58-fc46-11e6-9658-9ab6294ca397.png)

**Output**

![image](https://cloud.githubusercontent.com/assets/550895/23344205/224fc58e-fc46-11e6-915d-c93cefb04012.png)

**Benchmark**

```bash
$ node bench.js
simple x 117,605 ops/sec ±0.57% (98 runs sampled)
```
CC: @Turfjs/ownership

**Comment by mourner** [comment]
Not a fan of bringing yet another computational geometry library as a dependency...

**Comment by DenisCarriere** [comment]
@mourner Thanks for your input, I totally agree with your comment. Recently I've sent over many commits to the library (https://github.com/MartyWallace/PolyK/commits/master) before including it as a dependency.

Mourner you are the expert in Javascript computational geo libraries (big fan).

I'm willing to "Turfify" this library to Zero dependency (only Turf dependencies).

@mourner Can you review it after it's done (~1-2 weeks).

**Comment by mourner** [comment]
@DenisCarriere oh, don't take my comment as the final word — it's just my first thought. Worth discussing. Additional questions that come to mind are:

- How much size does PolyK add to the bundle?
- How stable/robust/fast is the algorithm?
- How much of an overlap there is between JSTS (already a dependency) and PolyK?
- Is it really necessary to have this as a core package, or would it be easier to just publish as a separate module and free to use PolyK/any other deps needed?

@morganherlocker @tcql thoughts?

**Comment by morganherlocker** [comment]
Same questions as @mourner for me, and one more. I definitely think this could be useful as a core module.

What should the behavior be when a line partially slices a polygon, but does not fully cleave it into 2+ parts? I ask because this is a bit ambiguous and one seemingly intuitive behavior (it inserts a new self-touching edge(s)) would lead to invalid geometries.

**Comment by DenisCarriere** [comment]
Ok I've got a first draft (80% done), I really like the idea of having this as a core module without even the JSTS dependency.

I've built the ~`@turf/slice`~ `@turf/polygon-slice` module entirely from TurfJS modules.

**TurfJS dependencies**
- `@turf/meta` (iterate over coordinates)
- `@turf/point-on-line` (Detects edges where to cut)
- `@turf/inside` (Helps detect if line is inside polygon)
- `@turf/helpers` (duh)

**Issues with TurfJS dependencies**
- `@turf/line-slice` Had to refactor this one a bit, had issues when trying to match the exact coordinate of a line, the slice always had an offset. Also need to find a way to do a reverse slice, the linestring giving has the same start & end coordinate.

**Next steps**
- Cleaving into 2+ parts would definitely be a challenge, I don't think it's impossible. If a line cleaves 1 or 2 or 3 parts, the output will always be a `FeatureCollection<Polygon>` with the parts inside the collection. If the linestring did not slice any parts, the polygon is simply returned as a `FeatureCollection<Polygon>`. I also I think the properties of the original polygon should translate down into the sliced pieces (no additional properties).
- Slicing polygons with holes (donuts), big challenge 👍 

**Performance**
As for performance, it's decent... (could always be faster).

```
complex x 2,752 ops/sec ±0.64% (98 runs sampled)
exact-coordinates x 2,825 ops/sec ±3.06% (92 runs sampled)
half-outside x 6,600 ops/sec ±0.67% (100 runs sampled)
outside x 17,133 ops/sec ±0.69% (98 runs sampled)
simple x 3,034 ops/sec ±0.99% (97 runs sampled)
```

**Debugging**

I've got some "fancy" looking debug geojson results.

https://github.com/Turfjs/turf/tree/slice/packages/turf-polygon-slice/test/out

**Debug mode**
![image](https://cloud.githubusercontent.com/assets/550895/23387743/7de1b222-fd2c-11e6-8e27-a1fc502629f6.png)

**Actual ouput**
![image](https://cloud.githubusercontent.com/assets/550895/23387767/a30a7020-fd2c-11e6-86a4-e5adb9f8be72.png)

CC: @mourner @morganherlocker

**Comment by rowanwins** [comment]
Nice one @DenisCarriere - looks like you're making good progress!

**Comment by DenisCarriere** [comment]
:) thanks, it's definitely shaping up, I might need to take a step back and fix/improve/test some of the turfjs modules I'm using as dependencies.

**Comment by DenisCarriere** [comment]
@rowanwins Ok this module is long overdue... I'm going to work on this today/week.

A few modules I've needed before to make this possible:

- [geojson-rbush](https://github.com/DenisCarriere/geojson-rbush) (GeoJSON implementation of [RBush](https://github.com/mourner/rbush))
- [@turf/line-segment](https://github.com/Turfjs/turf/tree/master/packages/turf-line-segment)
- [@turf/line-intersect](https://github.com/Turfjs/turf/tree/master/packages/turf-line-intersect)

**Comment by erikh2000** [comment]
@DenisCarriere I just got done doing the "new project" setup stuff, and am digging into polygonSlice(). A thing that confuses me with the current implementation is that it returns a FeatureCollection of Linestrings. I thought it was supposed to return sliced polygons. Is that just "work in progress" status of the polygonSlice() function? Or maybe I'm confused about how the function is intended to work.

**Comment by DenisCarriere** [comment]
Polygon Slice should return a FeatureCollection of Polygons, it might of still been in Debugging mode.

In reality Polygons are simply just a collection of connected lines, only difference is the Inner & Outer rings which can be [determined by the Winding](https://macwright.org/2015/03/23/geojson-second-bite.html#winding).

The [`slice branch`](https://github.com/Turfjs/turf/tree/slice/packages/turf-polygon-slice) would be considered a "work in progress" (aka: full rewrite).

Ref: 
- ~~https://github.com/Turfjs/turf-is-clockwise~~ =>  [`@turf/boolean-clockwise`](https://github.com/Turfjs/turf/tree/master/packages/turf-boolean-clockwise)
- https://macwright.org/2015/03/23/geojson-second-bite.html#winding

**Comment by erikh2000** [comment]
Okay, thanks, that makes sense. I just wanted to make sure I wasn't missing something in what the function was supposed to do.

**Comment by stebogit** [comment]
Just as an update note, [`@turf/boolean-clockwise`](https://github.com/Turfjs/turf/tree/master/packages/turf-boolean-clockwise) is now integrated as TurfJS module

**Comment by alexgleith** [comment]
Hey @DenisCarriere, I'm looking at this today, as I did have it working, but the `@turf/polygonize` function is no longer returning two features for the split... it seems to be cleverly dissolving the split line component.

So, I wonder if you have any thoughts about a way to go about re-uniting the lines that are output currently into polygon features?

Another thing, I think it should work for multi-polygon features. Not sure the best way to do that, but though having another path with some recursion would sort it out?

**Comment by alexgleith** [comment]
Hey @DenisCarriere 

One more note. I realised that the issue causing the polygonize to fail is in line splitting. In the `simple` example in the test cases, the lines that are split up the top of the feature split differently depending on if you split the line with the polygon or the polygon with the line.

Evidence 1 - this is the difference between the split points:
![image](https://user-images.githubusercontent.com/3445853/31060720-d3620a9e-a763-11e7-8b36-8a550203e03b.png)

Evidence 2 - this is the output of polygonize with the bad topology:
![image](https://user-images.githubusercontent.com/3445853/31060727-feb54332-a763-11e7-8b4e-f71b9b87a61d.png)

Evidence 3 - this is the output of polygonize after doing a `truncate` on the split features:
![image](https://user-images.githubusercontent.com/3445853/31060737-20892186-a764-11e7-80ef-3c9c808315f1.png)

In running through the tests, I think my fix is a bad workaround to what is perhaps an issue in the line-splitter. I might put together a test case and a bug report.

**Comment by DenisCarriere** [comment]
@alexgleith Great job!

> it seems to be cleverly dissolving the split line component.

🤔 That's annoying and good at the same time...

> I think it should work for multi-polygon features.

👍 Agreed, however getting Polygon to work first would be the first step.

> perhaps an issue in the line-splitter

🤔 Definitely submit a test case for the line splitter, we can look into fixing that first.

**Comment by alexgleith** [comment]
> Definitely submit a test case for the line splitter, we can look into fixing that first.

Ok, I'll do that now.

I have some working code for handling MultiPolygons. Not sure if it fits into the Turf 'standard way of doing things' but it works. Will leave it for now, as it's pretty trivial, but I can put a PR together if you like.

**Comment by alexgleith** [comment]
Hey @DenisCarriere, here's an [issue](https://github.com/Turfjs/turf/issues/990) and it links a [PR](https://github.com/Turfjs/turf/pull/989) with a test case.

**Comment by DenisCarriere** [comment]
Thanks @alexgleith for submitting this, this will be useful as reference whenever someone tackles this issue.

At the moment I'm not too sure where to start with this issue. 🤷‍♂️

**Comment by skaletech** [comment]
We've added a function that may be of interest based on the original Turf-cut found here: https://github.com/Turfjs/turf-cut

Here it is: https://github.com/skaletech/turf-cut

It works by creating a small buffer around the linestring, then calling turf.difference, then zipping points near the line to the line, then filtering out duplicate points.  This may not work for every use case, but it works well in our use case where points are usually much more distant than the tolerance input.

Also,  one small improvement would be to test whether a point was in the original polygon before allowing it to be 'zipped'.

**Comment by DenisCarriere** [comment]
@skaletech Cool work! Seems like a nice hack around the solution, unfortunately I don't think we would adopt this approach for the Polygon Slice, but it's very cool to see that this works for your use case. 👍

**Comment by skaletech** [comment]
Yep, we needed to allow for multiple line segments and self intersection.

So for example:

<img width="357" alt="before" src="https://user-images.githubusercontent.com/33358776/32446018-2d9f8ec6-c2cd-11e7-86d7-4d894d41d557.png">
<img width="357" alt="after" src="https://user-images.githubusercontent.com/33358776/32446025-37443238-c2cd-11e7-8290-0e4c0f356cb9.png">

We'd love to get a non-hack version if anyone has one.

**Comment by gdrouet** [comment]
Looking for this feature! It would be useful to split polygons that are too big.

**Comment by sroettering** [comment]
Any update on this?

**Comment by alexgleith** [comment]
I've had another look at the code, and ran it with the tests, and I have found that there are some issues.

I'm actually using it in production, with a couple of changes to the code here, including using `polygonize` on the output before it's returned and `truncate` on each of the features before they're polygonized. But with the test data, there are issues due to coordinate precision, where the polygonize process either fails or merges the split polygons back together! It's kind of hard to solve neatlly...

**Comment by rowanwins** [comment]
Gday @alexgleith 

Can you put up a pull request to a new branch with your version and I should be able to take a look. It would be great to get this module going because people have already put alot of effort it, it's just that final push to get it over the line!

**Comment by rowanwins** [comment]
PS also dropping a note here about [this blog post](https://geidav.wordpress.com/2015/03/21/splitting-an-arbitrary-polygon-by-a-line/) and [associated repo](https://github.com/geidav/concave-poly-splitter) 

PPS which also reminds me that it would be cool to have a `isConcave()` module - if you know a poly is concave then you can use that information to determine which algorithm you might use....

**Comment by alexgleith** [comment]
Hey @rowanwins 

My changes aren't significant, but here they are: https://github.com/alexgleith/turf/tree/slice

This commit captures it: https://github.com/alexgleith/turf/commit/1a47118892b5318bcc16aa3022a07c58b8d572b1

I really do think that [this issue](https://github.com/Turfjs/turf/issues/990) is part of the cause... I don't know how to fix it! Maybe implementing a more complex algorithm (as you've linked to) is the way to go.

**Comment by rowanwins** [comment]
Thanks for those links @alexgleith and your work on the module, will try and take a look this evening and see if I can find a clever way around it.

**Comment by daudihusbands** [comment]
@alexgleith I downloaded from your repo but I don't know how to build it. Sorry, I'm a newb to this stuff.

**Comment by rowanwins** [comment]
Hi @daudihusbands 

You might be able to download @alexgleith repo - then 
````
cd turf (where you downloaded the repo to)
npm install
rollup
```` 
This should generate a `main.js` file containing the bundled up version of Turf. The rollup command might require you to have rollup installed globally (eg `npm install rollup -g`). Hope that helps.

**Comment by rowanwins** [comment]
FYI I did look into this a little while back and the impression I got was that there would probably be benefit in moving to a more known algorithm like the one suggested above

**Comment by daudihusbands** [comment]
Thank you for your speedy response. Are you referring to the algorithm by @skaletech?

**Comment by alexgleith** [comment]
Hey @daudihusbands, the code I supplied does not work in all cases, so beware! It does work for most of my cases, though, and I'm using it in production.

**Comment by daudihusbands** [comment]
Thanks for the heads up @alexgleith. I am testing it out now. I'll post feedback if I run into any issues. Really appreciate your help. BTW, I didn't get rollup to work. I did notice that the config file is referring to index.js but I didn't find that file. No worries though, I copied the your code into my file.

**Comment by IsabellaCHY** [comment]
When is this module updated?  Is there a better solution?  This method is urgently needed now。

**Comment by avinashbitra** [comment]
Any timeline in when will this feature gets released ?

**Comment by morganherlocker** [comment]
> When is this module updated? Is there a better solution? This method is urgently needed now。

> Any timeline in when will this feature gets released ?

I am not aware of anyone working on this algorithm at the moment. The proposed solutions so far do not appear robust enough to make it to production. I would suggest using a lower level module or writing one yourself, and if that module works well, we can add it as a dependency in turf with a normalized interface.

**Comment by jacob-midtlyng-wsg** [comment]
Correct me if I'm wrong, but it seems like this feature was never added to Turf. We're simple in our Turf use so we updated the apparently abandoned [Turf/turf-cut](https://github.com/Turfjs/turf-cut) to work with the latest version of Turf and converted the module into a plain function. It's less elegant than @skaletech's solution above which you should review as well.

If it helps anyone who ends up here like we did, here's a link:
https://github.com/woodland-solutions-group/turf-cut

Since it's a hack I assume we should stop there. Please let me know if otherwise and we can try jumping the hoops to add it as a Turf module.

**Comment by eric-g-97477** [comment]
I too would like to see a supported way in turf to split a polygon instead of relying on another library.

Any chance this could happen?

**Comment by sasos90** [comment]
Still nothing i suppose? Looks like this feature is more complex than I thought.
